README.rst

```diff
@@ -308,6 +308,122 @@ In case you need to run additional instructions just before the build step you c
 
 You can find more patches in the `patch catalog <#template-patch-catalog>`_ below.
 
+Using Frontend Plugin Slots
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``mfe-dockerfile-pre-npm-build`` patch is also suitable for taking advantage of the ``env.config.jsx`` mechanism, which lets you define and use frontend plugins.  For instance, you can use it to modify the footer across all MFEs *without* having to maintain a fork of ``frontend-component-footer``.
+
+Start by following the Tutor plugin tutorial at the `Adding new templates <https://docs.tutor.edly.io/tutorials/plugin.html#adding-new-templates>`_ section.  Instead of adding a whole new ``Dockerfile``, though, you'll just modify the exiting one.  Your ``myplugin.py`` should look something like this:
+
+.. code-block:: python
+
+    import os
+    from tutor import hooks
+
+    # Add the template root folder
+    template_folder = os.path.join(os.path.dirname(__file__), "templates")
+    hooks.Filters.ENV_TEMPLATE_ROOTS.add_item(template_folder)
+
+    # env.config.jsx files should live in `mfe/build/mfe`, which is also where the
+    # tutor-mfe Dockerfile goes.  This is so it's more straightforward to copy it
+    # in, as seen below.
+    hooks.Filters.ENV_TEMPLATE_TARGETS.add_item(("mfe/build/mfe", "plugins"))
```

Comment:
I just realized I didn't address one major point that @hinakhadim hammered home afterwards.  That multiple plugins may want to modify env.config.jsx.  I have to think some more about this.

Reply:
I just went over [Indigo's `env.config.jsx`](https://github.com/overhangio/tutor-indigo/blob/nightly/tutorindigo/templates/indigo/env.config.jsx), and I guess that's as good an example as any of my first point above. :)

The question I'm ruminating is: how do we let plugins do all that in a single file and still allow multiple plugins to coexist there?

Maybe the answer is that we should _not_ allow plugin authors to run wild with `env.config.jsx`.  Maybe all they can do is npm-install a plugin package via the `post-npm-install` patch,  import the components via a new `mfe-env-config-imports` patch, then declare them via `mfe-env-config-plugin-slots`.

Reply:
Multiple plugins modifying `env.config.jsx` is definitely a tricky problem, and the fact that [`tutor-indigo` already leverages it](https://github.com/overhangio/tutor-indigo/blob/nightly/tutorindigo/templates/indigo/env.config.jsx) makes it quite the pressing issue.

The first thought that came to my mind was providing a way for plugins to "add their changes to the pile" so to speak. I tested the following `env.config.jsx` file

```jsx
import { DIRECT_PLUGIN, PLUGIN_OPERATIONS } from '@openedx/frontend-plugin-framework';

const config = {
  pluginSlots: {
    desktop_header_slot: {
      keepDefault: false,
      plugins: [
        {
          op: PLUGIN_OPERATIONS.Insert,
          widget: {
            id: 'custom_header_component_one',
            type: DIRECT_PLUGIN,
            RenderWidget: () => (
              <h1>1</h1>
            ),
          },
        },
      ]
    },
    desktop_header_slot: {
        keepDefault: false,
        plugins: [
        {
          op: PLUGIN_OPERATIONS.Insert,
          widget: {
            id: 'custom_header_component_two',
            type: DIRECT_PLUGIN,
            RenderWidget: () => (
              <h1>2</h1>
            ),
          },
        },
      ]
    },
  },
}

export default config;
```

and found that FPF handles it somewhat gracefully - the second component (the one that renders `2`) is shown and the first component (the one that renders `1`) is not.

In a world where each tutor plugin *only* touches one FPF slot this could be helpful - we could provide a way to "add changes to the pile" and have operators make sure the plugins are doing so in their preferred order. The obvious problem with that, however, is tutor plugins that utilize multiple FPF slots. If plugins A and B both replace the header and footer, and a site operator wants the header from A and footer from B this falls apart.

I'll keep thinking about this. It is definitely not a simple problem.

Reply:
Thanks for the productive discussion and @brian-smith-tcril  for testing.

I’m currently facing a challenge with the scenario you mentioned: "If plugins A and B both replace the header and footer, and a site operator wants the header from A and the footer from B, this falls apart."

There was a similar instance where a user needed to use both tutor-indigo and a second plugin for additional customizations. From what I understand, the priority of each plugin’s components is determined by the order of activation. If tutor-indigo is enabled first, followed by the second plugin, the components from the second plugin will have priority.
```Dockerfile
...
RUN npm install @edx/frontend-component-header@npm:indigo-header
RUN npm install @edx/frontend-component-footer@npm:indigo-footer

RUN npm install @edx/frontend-component-header@npm: custom-header
RUN npm install @edx/frontend-component-footer@npm: custom-footer
...
```

 Conversely, if the second plugin is enabled first, tutor-indigo components will take precedence (means we use indigo header/footer in react js project).
```Dockerfile
...
RUN npm install @edx/frontend-component-header@npm: custom-header
RUN npm install @edx/frontend-component-footer@npm: custom-footer

RUN npm install @edx/frontend-component-header@npm: indigo-header
RUN npm install @edx/frontend-component-footer@npm: indigo-footer
...
```

In cases where the operator needs the header from tutor-indigo (assuming that user cannot alter plugin) and the footer from the custom plugin (Plugin B), they would need to adjust Plugin B to exclude its header component, keeping only the footer. This way, Plugin A (Indigo) provides the header, and Plugin B supplies the footer without conflicts.

```Dockerfile
RUN npm install @edx/frontend-component-header@npm: indigo-header
RUN npm install @edx/frontend-component-footer@npm: indigo-footer

RUN npm install @edx/frontend-component-footer@npm: custom-footer
```

Any further insights would be greatly appreciated!

Reply:
> I’m currently facing a challenge with the scenario you mentioned: "If plugins A and B both replace the header and footer, and a site operator wants the header from A and the footer from B, this falls apart."

It is my understanding that this scenario is handled fully in https://github.com/overhangio/tutor-mfe/pull/234